### PR TITLE
Another containerd instance to run after vault open

### PR DESCRIPTION
**Pull request by giggsoff**
To ensure encryption we can run the second containerd instance after vault open.

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:157]: Instead of waiting I think we need to have domainmgr react to VaultStatus changes and when the vault is unlocked, it will start the user containerd. Thus the upgradeconverter will just to the moving of the files.

Otherwise domainmgr will not progress to report AssignableAdapters etc until the vault has been unlocked.
Also, it will be unable to start domains whose volumes live in /persist/clear/  before the vault is unlocked. (I don't think EVE-OS currently does that due to how volumemgr works, but we shouldn't add more obstacles to making it work.)


In addition, one thing we didn't talk about on the call is how containerd will

Comment (giggsoff) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:157]: Good point, will modify.
It seems your message ended in the half of sentence for some reason.

**Comment by giggsoff** [comment]
Made some research around including containerd as a code in our modules, seems it is quite difficult:

1. Containerd wants to use boltdb for metadata and db should be opened from one place, but we have calls to CAS in different agents, so still need separate process (agent) to handle connection.
2. Not all methods we use are available without Client (grpc connection), for example [image](https://github.com/containerd/containerd/blob/main/image.go). We will be forced to re-implement them.

**Comment by deitch** [comment]
> Made some research around including containerd as a code in our modules, seems it is quite difficult

containerd includes lots of code that can be used as a library, but it wasn't really built for that. At heart, it is a daemon service that is intended to be run as such.

Thinking about this a bit more, it only really runs "as a service" when a client connects to it and sends a command. It then does various things around images in its store and calling a container runtime (like runc, but could be others). If the entire set of clients is under our control, and that which is under control is written in go and capable of including the functionality as a library, then we could bypass the additional running service.

To try and clarify this:

The normal flow would be:

```
  client ---> containerd (daemon) ---> make changes to images in store, download, etc.  ---> return
                                                     |-----> start/stop/etc. containers (usually using runc)        ---> return
```

Between the end of `return` in each phase, and the next call of `client ---> containerd`, it is sitting around doing nothing.

If the entire set of clients is just one process, which we control, then we _could_ do the following: 

```
  client ---> make changes to images in store, download, etc.
        |-----> start/stop/etc. containers (usually using runc)
```

It would put additional burden on us, e.g. we would have to be _positive_ no other thread or process is accessing these containers or image store at the same time. That part is easy - just have a singleton struct that controls access.

I am reasonably confident we _could_ do it. On the image side, linuxkit does precisely this, managing images on its own without containerd. I would have the following hesitations about whether we _should_:

* The learning curve for engineering eve gets a bit harder, where it already is a pretty steep curve.
* I am not 100% certain that everything we need from containerd has a proper library surface; we might just hit up against things that would not be exposed, and thus be blocked. I find this unlikely, but possible.

My recommendation would be to continue the current approach and get it working. If that meets our needs, and the only issue is the process overhead (files already exist on the system, so the only overhead should be some memory from the go runtime), then we can approach this separately.

File: pkg/pillar/Dockerfile
Comment (deitch) [inline pkg/pillar/Dockerfile:3]: Nothing wrong with moving to a proper release in pillar, but let's not mix these two. This and its dependencies should be its own PR.

File: pkg/pillar/build.yml
Comment (deitch) [inline pkg/pillar/build.yml:19]: We will need `ctr` as well, and `runc` too.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (deitch) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:558]: Does this mean that it is `domainmgr` that starts the user containerd? Is that the right place to start it?

File: pkg/pillar/containerd/containerd.go
Comment (deitch) [inline pkg/pillar/containerd/containerd.go:48]: If I understood correctly, then inside pillar, there is only one containerd ever to access, the user one. The system one is the one that runs pillar. Pillar (I think) would never have reason to access the system one. So it won't be mounted in or accessible. Could we not then make this just `ctrdSocket = "/run/containerd/containerd.sock"` (which is the user one, but as far as pillar is concerned, the only one)? 

That also would change the `waitForUserContainerd()` to just `waitForContainerd()`. It also would simplify many of these changes, as they still would just look the same: start/stop/remove/etc containers, and load/unload images, all in the single containerd that pillar knows about.

Is there a reason that pillar still needs access to the system one?

File: pkg/pillar/hypervisor/containerd.go
Comment (deitch) [inline pkg/pillar/hypervisor/containerd.go:29]: Same for here. Should this not be in the _user_ containerd, i.e. the only one that really exists as far as pillar is concerned?

File: pkg/pillar/containerd/containerd.go
Comment (giggsoff) [reply pkg/pillar/containerd/containerd.go:48]: Right now user containerd is here to be used by CAS (download/prepare data in /persist/vault/volumes). System containerd still here to run domains

Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: Is there any reason to continue to do so? From the pillar code's perspective, containerd is containerd. It shouldn't care. If you already are segregating (and I still think this is a good approach you have), then might as well do it completely.

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (giggsoff) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:558]: Any suggestions? I received https://github.com/lf-edge/eve/pull/2524#discussion_r811426341

File: pkg/pillar/build.yml
Comment (giggsoff) [reply pkg/pillar/build.yml:19]: I use system containerd to run domains, so no needed

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (deitch) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: It needs to be started at the beginning of pillar. I do not know what the right location is though.

File: pkg/pillar/build.yml
Comment (deitch) [reply pkg/pillar/build.yml:19]: We still need the binaries.

File: pkg/pillar/containerd/containerd.go
Comment (giggsoff) [reply pkg/pillar/containerd/containerd.go:48]: We want to communicate with xen-tools running in system containerd in hypervisor. And in general run domains based on xen-tools we build inside linuxkit (system containerd).

File: pkg/pillar/cmd/domainmgr/domainmgr.go
Comment (giggsoff) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: The main reason to have it (the second containerd) is to run after vault unlocked to use decrypted directories.

File: pkg/pillar/containerd/containerd.go
Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: Isn't hypervisor control inside pillar as well? You started segregation, might as well go all the way, unless there is a reason we cannot.

Comment (giggsoff) [reply pkg/pillar/containerd/containerd.go:48]: The problem I can see: hypervisor module for xen wants to talk with "xen-tools" container which is inside "system" containerd in many places ([for example](https://github.com/lf-edge/eve/blob/e2c144539ea2dfbb820e19f8c612b84e9166a768/pkg/pillar/hypervisor/xen.go#L569) when we work with AssignableAdapters) and as @eriknordmark mentioned https://github.com/lf-edge/eve/pull/2524#discussion_r811426341 we want to manage adapters before/in parallel with vault unlock, i.e. before start of the second containerd.

Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: > hypervisor module for xen wants to talk with "xen-tools" container which is inside "system" containerd in many places

I am not sure I understand this. Looking at the example to which you linked, I think you are saying that:

1. `xen-tools` runs as a standalone system-level container, parallel to pillar (see [here](https://github.com/lf-edge/eve/blob/1c8a58d227ab955639273552771ff5f1eae2ab5f/images/rootfs.yml.in#L61-L64))
2. `hypervisor` module (which is part of `pillar`) needs to do things to various VM instances
3. `hypervisor` does so by exec-ing into the running system-level container
4. Therefore `hypervisor` needs to be have access to the system-level containerd, so it can do the exec-into tasks

Is that it? If so, is it only hypervisor and its xen-tools requirements? Or are there other areas that also need access to system-level?

> as @eriknordmark mentioned https://github.com/lf-edge/eve/pull/2524#discussion_r811426341 we want to manage adapters before/in parallel with vault unlock, i.e. before start of the second containerd.

I don't understand. What does this have to do with system vs workload containerd?

Comment (giggsoff) [reply pkg/pillar/containerd/containerd.go:48]: > Is that it? If so, is it only hypervisor and its xen-tools requirements? Or are there other areas that also need access to system-level?

Yes, and for this reason I use system containerd right now.

> I don't understand. What does this have to do with system vs workload containerd?

I just mention expected sequence, we want to interact with system containerd before vault unlocked and we want to run user containerd after vault unlocked.

Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: Looking into it, I see that `CtrSystemExec()` is called 7 times in [xen.go](https://github.com/lf-edge/eve/blob/e2c144539ea2dfbb820e19f8c612b84e9166a768/pkg/pillar/hypervisor/xen.go):

* `xl dmesg` - 2x
* `xl destroy` - 1x
* `xl pci-assignable-add` - 1x
* `xl pci-assignable-rem` - 1x
* `xl info` - 1x
* `xentop -b -d 1 -I 2 -f` - 1x

Is that the total of it?

Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: > in general we use /containers/services/xen-tools prepared by linuxkit to start our domains

Do you mean other than the above 7 places? I didn't see anything there to start a domain.

I should remember all of this. 🤷‍♂️

Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: Oh right, [here](https://github.com/lf-edge/eve/blob/e2c144539ea2dfbb820e19f8c612b84e9166a768/pkg/pillar/hypervisor/xen.go#L131-L155). That it?

File: pkg/dom0-ztools/rootfs/etc/containerd/config.toml
Comment (eriknordmark) [inline pkg/dom0-ztools/rootfs/etc/containerd/config.toml:2]: How much disk space will we use in the system root? Need to keep in mind that this consumes memory.

File: pkg/pillar/utils/waitfor.go
Comment (eriknordmark) [inline pkg/pillar/utils/waitfor.go:158]: Probably makes sense for this to be log.Noticef so we can see if any microservice is stuck due to this.

**Review by eriknordmark — APPROVED** [review]
I see there are still some open comments/questions from @deitch 

Also, shouldn't we document this somewhere? (Do we already have some file describing how we interact with containerd?)

Kicking off the tests to see how this does.

File: pkg/dom0-ztools/rootfs/etc/containerd/config.toml
Comment (giggsoff) [reply pkg/dom0-ztools/rootfs/etc/containerd/config.toml]: Good point, thank you. I set it to another directory in persist to not consume memory.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by deitch** [comment]
> I see there are still some open comments/questions from @deitch

been a bit slow to get back to this, looking now

File: pkg/pillar/containerd/containerd.go
Comment (deitch) [reply pkg/pillar/containerd/containerd.go:48]: Having had a chance to think this one through, I still agree with your approach. I also think we should do complete segregation: user containerd inside pillar to start userspace both VM domains and containers, system containerd to start system-level services, which pillar will not even be aware of. FWIW, rancheros was doing this years ago, and docker for Mac/windows does something a bit similar.

That having been said, we should not hold this up for that. We should get this in, then put in the effort to figure out how to get all of the userspace workloads entirely separate from system

**Comment by deitch** [comment]
Addressed all of them. The core really is about what should be in user containerd and what in system containerd. We are on the right path, just haven't walked all the way. We should not hold up though ("perfect enemy of good" and all that).

**Comment by eriknordmark** [comment]
> Addressed all of them. The core really is about what should be in user containerd and what in system containerd. We are on the right path, just haven't walked all the way. We should not hold up though ("perfect enemy of good" and all that).

@deitch are there changes you think we should do before merging this PR, or do you think we can merge now and refine down the road?

**Review by eriknordmark — APPROVED** [review]
Run eden again

**Comment by eriknordmark** [comment]
FWIW this is passing ztests. @petr-zededa are you planning some changes?

**Comment by giggsoff** [comment]
> FWIW this is passing ztests. @petr-zededa are you planning some changes?

Not every retry as I can see, will check logs.

**Comment by eriknordmark** [comment]
> > FWIW this is passing ztests. @petr-zededa are you planning some changes?
> 
> Not every retry as I can see, will check logs.

ztests is flakier than before on the packet systems so it takes several tries to a PR to pass.

**Comment by giggsoff** [comment]
> > > FWIW this is passing ztests. @petr-zededa are you planning some changes?
> > 
> > 
> > Not every retry as I can see, will check logs.
> 
> ztests is flakier than before on the packet systems so it takes several tries to a PR to pass.

Yes, cannot find oddities in logs. I slightly changed cgroups memory limit logic to properly add 50 megabytes overhead.

**Pull request merged**
